### ISSUE TITLE
cost diff not detected

### ISSUE THREAD

**Issue by jrobison-sb** [infracost/infracost#540]
This might be user error, but I'm not getting cost diffs as I would expect them. I have InfraCost running via the Atlantis integration (`infracost/infracost-atlantis:latest`, which is `v0.8.2` at the time of writing this). I populated an environment variable for the API key and for the Terraform binary. When I make a pull request Atlantis runs as expected and comments on the PR with a `plan`, and then InfraCost runs as expected but it reports a 0% cost diff. Here's an example of my `terraform plan` and also the output of infracost debug mode:

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
+ create

Terraform will perform the following actions:

  # module.production_application.aws_instance.just_testing will be created
+ resource "aws_instance" "just_testing" {
      + id                           = (known after apply)
      + instance_type                = "t3.large"
...
}

Plan: 1 to add, 0 to change, 0 to destroy.

#####
Running Infracost atlantis_diff.sh in debug mode, remove atlantis_debug=true from Atlantis configs to disable debug outputs.

Running infracost breakdown using:
  $ infracost breakdown --no-color --path /home/atlantis/.atlantis/repos/myorg/devops/2560/default/terraform/myorg/environments/production/production-default.tfplan --format json
time="2021-03-30T21:00:39Z" level=info msg="Detected Terraform plan file at /home/atlantis/.atlantis/repos/myorg/devops/2560/default/terraform/myorg/environments/production/production-default.tfplan"
time="2021-03-30T21:00:43Z" level=info msg="starting: Running terraform show"
time="2021-03-30T21:00:43Z" level=info msg="Running command: /usr/local/bin/terraform0.13.6 show -no-color -json production-default.tfplan"
time="2021-03-30T21:02:47Z" level=info msg="starting: Calculating monthly cost estimate"
Running infracost output using:
  $ infracost output --no-color --format diff --path infracost_breakdown.json
Comment not posted as diff (0%) is less than or equal to percentage threshold (0%).
```

How should I troubleshoot this? Thanks.

**Comment by alikhajeh1** [comment]
@jrobison-sb thanks for reporting this, can you please try the same thing with the env `INFRACOST_LOG_LEVEL=debug`? Maybe the CLI is not able to find a price for the instance, setting this env should show warnings if that's the case. Another way to debug is to share the TF resource block of `module.production_application.aws_instance.just_testing` with us so we can try to reproduce it.

**Comment by jrobison-sb** [comment]
@alikhajeh1 Thanks for your reply. I turned on `INFRACOST_LOG_LEVEL=debug` and was able to reproduce this. The output that was commented back to the pull request is massive and contains a good bit of our Terraform state, so I'll need to redact a lot of it. But here is what I can paste:

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
+ create

Terraform will perform the following actions:

  # module.production_application.aws_instance.just_testing will be created
+ resource "aws_instance" "just_testing" {
      + id                           = (known after apply)
      + instance_type                = "t3.large"
...
Plan: 1 to add, 0 to change, 0 to destroy.

#####
Running Infracost atlantis_diff.sh in debug mode, remove atlantis_debug=true from Atlantis configs to disable debug outputs.

Running infracost breakdown using:
  $ infracost breakdown --no-color --path /home/atlantis/.atlantis/repos/myorg/devops/2560/default/terraform/myorg/environments/production/production-default.tfplan --format json
time="2021-04-01T19:44:10Z" level=info msg="Detected Terraform plan file at /home/atlantis/.atlantis/repos/myorg/devops/2560/default/terraform/myorg/environments/production/production-default.tfplan"
time="2021-04-01T19:44:13Z" level=info msg="starting: Running terraform show"
time="2021-04-01T19:44:13Z" level=info msg="Running command: /usr/local/bin/terraform0.13.6 show -no-color -json production-default.tfplan"
time="2021-04-01T19:46:03Z" level=debug binary=terraform msg="{\"format_version\":\"0.1\",\"terraform_version\":\"0.13.6\",\"planned_values\":{\"root_module\":{\"child_modules\":[{\"resources\":[{\"address\":\"module.production_application...
...

# A few thousand lines of escaped json representing what looks to be most or all of our Terraform state, which I will redact

...

# Then a bunch of lines which look like this, one for each of the resources in the Terraform state, or at least one for each of the resources that InfraCost supports...

time="2021-04-01T19:46:16Z" level=debug msg="Getting pricing details from https://pricing.api.infracost.io/graphql for module.production_application.aws_s3_bucket.archive"

# And then a few dozen lines which repeat like this over and over...

time="2021-04-01T19:46:16Z" level=debug msg="diffResourcesByKey nil current and past with key "

Running infracost output using:
  $ infracost output --no-color --format diff --path infracost_breakdown.json
Comment not posted as diff (0%) is less than or equal to percentage threshold (0%).
```

The Terraform resource that is proposed in the pull request and for which I'm expecting a cost diff is mostly just a copy/paste from Terraform's [documentation](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance) and looks like this:

```
resource "aws_instance" "just_testing" {
  ami           = "ami-abcdef12345"
  instance_type = "t3.large"

  tags = {
    Name = "HelloWorld"
  }
}
```

Thanks for any thoughts you might have.

**Comment by alikhajeh1** [comment]
Thanks @jrobison-sb! That ruled-out the "not able to find a price" theory.

@aliscott / @sinabakh notice the `diffResourcesByKey` debug line above and how the [resourceKey](https://github.com/infracost/infracost/blob/master/internal/schema/diff.go#L61-L63) is not output in that debug line? any ideas?

**Comment by aliscott** [comment]
Not seeing anything immediately obvious. I wonder if it's something to do with how the module is structured. @jrobison-sb are you able to run the following with JQ so we can get an idea of the structure of the plan file:

```
terraform show -json plan.tfplan | jq '[paths as $path | select(getpath($path) | strings | contains("aws_instance.just_testing")) | $path]'
```

**Comment by jrobison-sb** [comment]
@aliscott Sure, thanks for investigating this.

```
[
  [
    "planned_values",
    "root_module",
    "child_modules",
    0,
    "resources",
    306,
    "address"
  ],
  [
    "resource_changes",
    306,
    "address"
  ],
  [
    "configuration",
    "root_module",
    "module_calls",
    "production_application",
    "module",
    "resources",
    329,
    "address"
  ]
]
```

**Comment by aliscott** [comment]
Thanks @jrobison-sb that looks expected and I haven't been able to reproduce with this structure yet.

Does the resource show when running either of these?
```
infracost diff --path=path/to/planfile.plan
```
or
```
infracost breakdown --path=path/to/planfile.plan
```

I wonder if the total cost is high and the cost change is minimal so we're rounding the percentage incorrectly to zero and therefore not showing a comment.

**Comment by jrobison-sb** [comment]
@aliscott Thanks, and after running those commands your hunch about rounding to zero is almost certainly correct. When using the planfile for my `aws_instance.just_testing` change, I can get `infracost breakdown` to show a breakdown of all the resources in my Terraform state which infracost supports, including the `just_testing` instance. And `infracost diff` is even more telling, in that it reports the ~$60 change for adding the instance and in the summary it does round to 0%.

Assuming that rounding to 0% is what's tripping me up here, and I'm pretty sure that it is, is that something I can unblock myself via configuring environment variables or something? I'd be happy enough if our pull requests could get a comment which said "this will change by $60" even if it didn't calculate a percentage diff of total monthly spend. Or is this something where it'll need a bugfix in the infracost application and then I'll need to pull a new Docker image after the bugfix is released?

Thanks for sticking with me on this.

**Comment by aliscott** [comment]
@jrobison-sb thanks. I'll push a fix to the latest Docker image in the next couple of hours to use 6dp precision for the calculations which I think should solve your issue.

Also, in the next release (probably tomorrow) we'll have this option `post_condition:'{"has_diff":"true"}'` for the script which will make sure it always posts a comment if there's a cost diff.

**Comment by jrobison-sb** [comment]
@aliscott thanks for fixing this. I see that https://github.com/infracost/infracost/pull/553 was merged and that separately 0.8.3 was released. A new image was pushed for infracost/infracost, but I don't see a corresponding new image for infracost/infracost-atlantis. Could you push a new image for Atlantis with this fix?

**Comment by alikhajeh1** [comment]
@jrobison-sb sorry for the delay, I had prepared all the PRs and got side-tracked, the infracost/infracost-atlantis image is now updated with the fix @aliscott mentioned as well as the new post_condition feature: https://github.com/infracost/infracost-atlantis/#post_condition
